In [138]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
import collections
from collections import defaultdict

# Loading the data

In [139]:
df_tweets = pd.read_csv("C:\\Users\\USER\\Documents\\Geetha\\Data Science\\Capstone Project 1\\tweets.csv", parse_dates=True)  

In [140]:
df_tweets = df_tweets.set_index(['time','name','username'])

In [141]:
df_tweets.head(20)


description  \
time            name          username                                                        
1/6/2015 21:07  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   

                                              location  followers  \
time            name          username                              
1/6/2015 21:07  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70      NaN        640   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70      NaN        640   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70      NaN        640   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70      NaN        640   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70      NaN        640   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70      NaN        640   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70      NaN        640   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70      NaN        640   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70      NaN        640   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70      NaN        640   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70      NaN        640   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70      NaN        640   

                                               numberstatuses  \
time            name          username                          
1/6/2015 21:07  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:51 

In [142]:
df_tweets.info()


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       14728 non-null object
location          11432 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


#### Columns 'description' and 'location' have null values

In [143]:
df_tweets['description'] = df_tweets['description'].fillna(' ')

In [144]:
df_tweets['location'] = df_tweets['location'].fillna(' ')

#### NaNs replaced with empty string in the columns 'description' and 'location' 

In [145]:
df_tweets.head(20)

description  \
time            name          username                                                        
1/6/2015 21:07  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   

                                              location  followers  \
time            name          username                              
1/6/2015 21:07  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70                 640   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70                 640   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70                 640   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70                 640   

                                               numberstatuses  \
time            name          username                          
1/6/2015 21:07  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:51 

In [146]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       17410 non-null object
location          17410 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


In [147]:
df_tweets=df_tweets.drop_duplicates()

In [148]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       17410 non-null object
location          17410 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


In [149]:
df_tweets.columns


Index(['description', 'location', 'followers', 'numberstatuses', 'tweets'], dtype='object')

In [150]:
df_tweets['description'] = df_tweets['description'].str.lower()
df_tweets['tweets'] = df_tweets['tweets'].str.lower()

In [151]:
df_tweets.head(20)


description  \
time            name          username                                                        
1/6/2015 21:07  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70  english translations: http://t.co/qldj0ftews   

                                              location  followers  \
time            name          username                              
1/6/2015 21:07  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70                 640   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70                 640   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70                 640   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70                 640   

                                               numberstatuses  \
time            name          username                          
1/6/2015 21:07  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:51 

In [152]:
df_tweets.index.get_level_values(1).value_counts(normalize=True)



Rami                    0.084721
War BreakingNews        0.068409
Conflict Reporter       0.062895
Salahuddin Ayubi        0.060655
Ibni Haneefah           0.040724
                          ...   
Khilafah Superstate     0.000402
Ibn Al-Shishani         0.000287
Abdul05                 0.000287
Newerajihadi61          0.000230
fahad slyman alsomal    0.000230
Name: name, Length: 112, dtype: float64

#### We are analyzing the tweets of 112 different users who have generated 17410 tweets

In [153]:

df_tweets['tweets'] =  df_tweets['tweets'] .apply(lambda elem: re.sub(r"(!|@|#|\n|')", "", elem)) 

#### Removing some of the special characters from 'tweets'column using regular expression operations

In [154]:
df_tweets['tweets'][6]



'english transcript : oh murabit : http://t.co/hujlj9kgkg http://t.co/t9ixmtbvgk'

In [155]:
df_tweets['tweets']

time             name              username       
1/6/2015 21:07   GunsandCoffee     GunsandCoffee70    english translation: a message to the truthful...
1/6/2015 21:27   GunsandCoffee     GunsandCoffee70    english translation: sheikh fatih al jawlani f...
1/6/2015 21:29   GunsandCoffee     GunsandCoffee70    english translation: first audio meeting with ...
1/6/2015 21:37   GunsandCoffee     GunsandCoffee70    english translation: sheikh nasir al wuhayshi ...
1/6/2015 21:45   GunsandCoffee     GunsandCoffee70    english translation: aqap: response to sheikh ...
                                                                            ...                        
5/13/2016 15:46  Anaksabil97       nvor85j            breakingislamicstate amaqagency fridaythe13th ...
5/13/2016 15:50  Anaksabil97       nvor85j            breaking iraqi forces bulldozer destroyed by g...
5/13/2016 15:51  Anaksabil97       nvor85j            breaking amaqagency islamicstate eastofhaditha...
5/13/2016 16:

In [156]:
'''

def df_tweets_tokenize(x):
    return x.str.split(' ').sum()
tweets_split= df_tweets.apply(df_tweets_tokenize(df_tweets['tweets']))   
'''

"\n\ndef df_tweets_tokenize(x):\n    return x.str.split(' ').sum()\ntweets_split= df_tweets.apply(df_tweets_tokenize(df_tweets['tweets']))   \n"

#### The tweets in 'tweets' column are splitted and are represened as a list in nw column 'tweets_split'

In [157]:
tweets_split= df_tweets['tweets'].str.split(' ').sum()
#print(tweets_split)
count_stopwords = collections.Counter(tweets_split)
#print(count_stopwords)

In [158]:

stopwords_morethan50 = dict()
for (key, value) in count_stopwords.items():
   # Check if key is even then add pair to new dictionary
   if value >= 50:
       stopwords_morethan50[key] = value
 

#print(stopwords_morethan50)




In [159]:
#tweets_http = df_tweets['tweets'].str.contains('https' ).sum()

In [160]:
#tweets_http

### Make a dictioary: Tokenize

In [161]:

list_words= ['turkey','islamicstate','syria','aleppo','warreporter1','russia', 'assad', 'iraq']
dict_token=defaultdict(list)
for value in list_words:
   for key in stopwords_morethan50:
      if value in key:           
           dict_token[value].append(key)
        
    
    

The words 'turkey','islamicstate','syria','aleppo','warreporter1','russia', 'assad', 'iraq' are found to be repeated in different formats

In [162]:
dict_token

defaultdict(list,
            {'turkey': ['turkey'],
             'islamicstate': ['islamicstate'],
             'syria': ['syria', 'syrian', 'syria:', 'syria.', 'syria|n'],
             'aleppo': ['aleppo', 'aleppo.'],
             'warreporter1': ['warreporter1:', 'warreporter1'],
             'russia': ['russia', 'russian', 'russia|n'],
             'assad': ['assad', 'assads'],
             'iraq': ['iraq', 'iraqi', 'iraq|i']})

#### Tf-idf

In [163]:
# Import TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer 
stop_words_list = ['english', 'is', 'with', 'by', 'oh', 'have', 'been', 'will', 'if', 'will', 'if',
                   'else', 'so', 'we', 'then', 'let', 'us', 'and', 'that', 'or', 'do', 'has', 'on', 'from', 'those', 'are', 
                   'their', 'were', 'etc', 'now', 'just', 'as', 'here', 'upon', 'into', 'also', 'at', 'its', 'while',
                    'had', 'youre', 'la', 'be', 'can', 'did', 'too', 'your', 'which']

# Instantiate TfidfVectorizer
tv = TfidfVectorizer(max_features=1000, stop_words=stop_words_list) 

# Fit the vectroizer and transform the data
tv_transformed = tv.fit_transform(df_tweets['tweets'] )

# Create a DataFrame with these features
tv_df = pd.DataFrame(tv_transformed.toarray(), 
                     columns=tv.get_feature_names()).add_prefix('TFIDF_')
tv.vocabulary_


{'message': 547,
 'to': 878,
 'the': 860,
 'in': 420,
 'syria': 831,
 'sheikh': 773,
 'abu': 28,
 'al': 55,
 'http': 408,
 'co': 208,
 'for': 331,
 'people': 644,
 'of': 607,
 'first': 325,
 'leader': 492,
 'aqap': 93,
 'victory': 927,
 'statement': 811,
 'it': 445,
 'second': 755,
 'series': 762,
 'soldier': 789,
 'jn': 457,
 'video': 928,
 'link': 506,
 'words': 964,
 'dawlah': 234,
 'please': 659,
 'share': 768,
 'our': 629,
 'new': 588,
 'account': 31,
 'after': 38,
 'previous': 672,
 'one': 618,
 'was': 942,
 'raid': 693,
 'france': 340,
 'taken': 835,
 'down': 266,
 'attack': 113,
 'mujahideen': 573,
 'https': 409,
 'saw': 748,
 'not': 598,
 'way': 944,
 'through': 874,
 'few': 315,
 'more': 563,
 'days': 236,
 'very': 924,
 'news': 589,
 'reports': 723,
 'syrian': 832,
 'should': 778,
 'still': 812,
 'path': 642,
 'watch': 943,
 'save': 747,
 'aleppo': 56,
 'released': 714,
 'group': 366,
 'islamic': 439,
 'no': 593,
 'rt': 733,
 'part': 640,
 'breaking': 163,
 'confirmed': 213,

In [164]:
# Isolate the row to be examined
sample_row = tv_df.iloc[0]

# Print the top 5 words of the sorted output
print(sample_row.sort_values(ascending=False).head(5))


TFIDF_http       0.637226
TFIDF_message    0.426213
TFIDF_sheikh     0.352366
TFIDF_abu        0.300994
TFIDF_co         0.220745
Name: 0, dtype: float64
